ansible/roles/atmo-ssh-setup/tasks/main.yml

```diff
@@ -1,145 +1,71 @@
-
-###
-# OTHER SSH CONFIG CHANGES
-##
-- name: Remove all AllowUsers lines to /etc/sshd_config
-  lineinfile: dest=/etc/ssh/sshd_config regexp="{{ item }}" state=absent
-  with_items:
-    - AllowUsers
-    - AllowGroups
-  tags:
-    - debug
-    - delete
-    
-- name: /etc/ssh/sshd_config changes
-  lineinfile: dest=/etc/ssh/sshd_config backup=yes regexp="{{ item.regexp }}" line="{{ item.line }}"
-  with_items:
-    - { regexp: '^PermitRootLogin', line: 'PermitRootLogin without-password'}
-    - { regexp: '^UseDNS', line: 'UseDNS no'}
-    - { regexp: '^PasswordAuthentication', line: 'PasswordAuthentication yes'}
-
-- name: Append AllowGroups Line to /etc/ssh/sshd_config 
-  lineinfile: dest=/etc/ssh/sshd_config line="AllowGroups {{ SSH_ALLOW_GROUPS }}"
-  tags:
-    - debug
-
-- name: manual OS detection
-  stat: path=/etc/redhat-release
-  register: os_ver
-  tags:
-    - debug
-    - ssh-restart
-
-- name: restart ssh if ubuntu
-  service: name=ssh state=restarted
-  when: not os_ver.stat.exists
-  tags:
-    - debug
-    - ssh-restart
-
-- name: restart sshd if centos
-  service: name=sshd state=restarted
-  when: os_ver.stat.exists
-  tags:
-    - debug
-    - ssh-restart
+# SSH as different users to determine which can connect (and determine linux distro)
+- block:
+  - name: Test default ansible connection and register root_connection
+    local_action: >
+      command ssh {{ SSH_OPTIONS }} -p {{ ansible_port }} root@{{ vm_ip }} exit
+    register: root_connection
+
+  - name: Test CentOS connection and register centos_connection
+    local_action: >
+      command ssh {{ SSH_OPTIONS }} -p {{ ansible_port }} centos@{{ vm_ip }} exit
+    register: centos_connection
+    when: root_connection.rc != 0
+
+  - name: Test Ubuntu connection and register ubuntu_connection
+    local_action: >
+      command ssh {{ SSH_OPTIONS }} -p {{ ansible_port }} ubuntu@{{ vm_ip }} exit
+    register: ubuntu_connection
+    when: root_connection.rc != 0 and centos_connection.rc != 0
+  ignore_errors: true
+- always:
+  - fail:
+      msg: "No remote connection established, atmo-ssh-setup can not continue."
+      when: use_remote_user is not defined
+
+# Set use_remote_user depending on which SSH task succeeded
+- block:
+  - name: Set use_remote_user variable to "centos"
+    set_fact: use_remote_user=centos
+    when: not centos_connection|skipped and centos_connection.rc == 0
+
+  - name: Set use_remote_user variable to "ubuntu"
+    set_fact: use_remote_user=ubuntu
+    when: not ubuntu_connection|skipped and ubuntu_connection.rc == 0
+
+  - name: Set use_remote_user variable to "root"
+    set_fact: use_remote_user=root
+    when: root_connection.rc == 0
+
+# Install Python2.7 for Ubuntu 16.04 since it uses Python3 by default and ansible_distribution
+# requires Python2.7. Ansible 2.2 includes experimental support for Python3
+- block:
+  - name: Manual Ubuntu version detection
```

Comment:
Small refactor, move this task outside the block, then add `when: use_remote_user == "ubuntu" && ubuntu16.rc == 0` as a clause to the block not each individual task. I'm not sure if the first part of the when is necessary.